--- icedtea6-1.8/Makefile.am.orig	2010-04-12 23:55:51.000000000 +0200
+++ icedtea6-1.8/Makefile.am	2010-04-16 19:40:57.496933541 +0200
@@ -408,11 +408,6 @@
 	patches/ecj/getannotation-cast.patch \
 	patches/ecj/override.patch
 
-if DTDTYPE_QNAME
-ICEDTEA_ECJ_PATCHES += \
-	patches/ecj/icedtea-jaxws-getdtdtype.patch
-endif
-
 ICEDTEA_ECJ_PATCHES += $(DISTRIBUTION_ECJ_PATCHES)
 
 # OpenJDK build environment.
